sql/wsrep_mysqld.cc

```diff
@@ -1665,86 +1666,82 @@ wsrep::key_array wsrep_prepare_keys_for_toi(const char *db, const char *table,
+                                    dd::String_type(table->table_name),
+                                    &table_obj)) {
         assert(0);
+        dd::release_mdl(thd, mdl_ticket);
+        continue;
       }
       collect_fk_children(table_obj, keys);
+      dd::release_mdl(thd, mdl_ticket);
     }
   }
+}
 
-  /* close the table and release MDL locks */
-  close_thread_tables(thd);
-  thd->mdl_context.rollback_to_savepoint(mdl_savepoint);
-  for (table = tables; table; table = table->next_local) {
-    table->table = NULL;
-    table->mdl_request.ticket = NULL;
+/*
+ * Collect wsrep keys corresponding to tables referenced by the given table
+ */
+static void collect_fk_parrents(const dd::Table *table_def,
+                                wsrep::key_array *keys) {
+  for (auto fk : table_def->foreign_keys()) {
+    keys->push_back(wsrep_prepare_key_for_toi(
+        fk->referenced_table_schema_name().c_str(),
+        fk->referenced_table_name().c_str(), wsrep::key::shared));
   }
 }
 
 void wsrep_append_fk_parent_table(THD *thd, TABLE_LIST *tables,
                                   wsrep::key_array *keys) {
   if (!WSREP(thd) || !WSREP_CLIENT(thd)) return;
-  TABLE_LIST *table;
-
-  thd->mdl_context.release_transactional_locks();
-  MDL_savepoint mdl_savepoint = thd->mdl_context.mdl_savepoint();
-
-  uint counter;
-  if (open_temporary_tables(thd, tables) ||
-      open_tables(thd, &tables, &counter,
-                  MYSQL_OPEN_FORCE_SHARED_HIGH_PRIO_MDL)) {
-    WSREP_DEBUG("unable to open table for FK checks for %s", thd->query().str);
-  }
-  for (table = tables; table; table = table->next_local) {
-    if (!is_temporary_table(table) && table->table) {
-      for (TABLE_SHARE_FOREIGN_KEY_INFO *fk = table->table->s->foreign_key;
-           fk < table->table->s->foreign_key + table->table->s->foreign_keys;
-           ++fk) {
-        WSREP_INFO("(p) appended fkey %s", fk->referenced_table_name.str);
-        keys->push_back(wsrep_prepare_key_for_toi(fk->referenced_table_db.str,
-                                                  fk->referenced_table_name.str,
-                                                  wsrep::key::shared));
+
+  for (auto *table = tables; table; table = table->next_local) {
+    if (!is_temporary_table(table)) {
+      MDL_ticket *mdl_ticket = nullptr;
+      if (dd::acquire_shared_table_mdl(thd, table->db, table->table_name, false,
+                                       &mdl_ticket)) {
+        assert(0);
+        continue;
       }
+
+      const dd::Table *table_obj = nullptr;
+      dd::cache::Dictionary_client::Auto_releaser releaser(thd->dd_client());
+      if (thd->dd_client()->acquire(dd::String_type(table->db),
+                                    dd::String_type(table->table_name),
+                                    &table_obj)) {
+        assert(0);
+        dd::release_mdl(thd, mdl_ticket);
+        continue;
+      }
+
+      collect_fk_parrents(table_obj, keys);
```

Comment:
Not an issue, but an improvement.

The implementation of functions `wsrep_append_fk_parent_table` and `wsrep_append_child_tables` are almost the same except the fact that the former calls `collect_fk_parrents()` and the latter calls `collect_fk_children()` in the end of the function.

I wonder if we could have a single function which accepts the callback (func_ptr or an instance of std::function), so that duplication of code can be reduced.

Reply:
Refactored using C++11 approach.